File: 152-maximum-product-subarray/maximum-product-subarray.java
class Solution {
    //krishna curmax and curmin are the continous max and min they dont stop okk
    //so we should not right like that okkk
    //u r good girllll!!!!!!!!
    public int maxProduct(int[] nums) {
        int max = Arrays.stream(nums).max().getAsInt();
        int curmax=1,curmin=1;

        for(int i=0;i<nums.length;i++){
            if(nums[i]==0){
                curmax=1;
                curmin=1;
                continue;
            }
            else if(nums[i]>0){
                curmax=Math.max(curmax*nums[i],nums[i]);
                curmin=Math.min(curmin*nums[i],nums[i]);
            }
            else{
               int tempmin=curmin;
               int tempmax=curmax;
                curmax=Math.max(tempmin*nums[i],nums[i]);
                curmin=Math.min(tempmax*nums[i],nums[i]);
            }

            if(curmax>max)
              max=curmax;
           
        }
        
        return max;
    }
}

File: 152-maximum-product-subarray/maximum-product-subarray.py
class Solution(object):
    def maxProduct(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """

        maxPro = 1
        minPro = 1
        ans = max(nums)

        for no in nums:
            if no == 0:
                maxPro = 1
                minPro = 1
                continue
            
            if no > 0:
                maxPro = max(maxPro * no, no)
                minPro = min(minPro * no, no)
            
            else:
                temp = maxPro
                maxPro = max(minPro * no, no)
                minPro = min(temp * no, no)
            
            ans = max(ans, maxPro)
        
        return ans


        